ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,765 @@
+        };
+        let validate_ctx = ctx.narrow(&Step::DZKPValidate).base_context();
+        let protocol_ctx = ctx.dzkp_upgrade(&Step::DZKPMaliciousProtocol, dzkp_batch);
+        Self {
+            batch_ref: batch_list,
+            protocol_ctx,
+            validate_ctx,
+        }
+    }
+}
+
+#[cfg(feature = "descriptive-gate")]
+impl<'a> Drop for MaliciousDZKPValidator<'a> {
+    fn drop(&mut self) {
+        self.is_safe().unwrap();
+    }
+}
+
+#[cfg(all(test, unit_test))]
+mod tests {
+    use std::iter::{repeat, zip};
+
+    use bitvec::{array::BitArray, order::Lsb0};
+    use futures::TryStreamExt;
+    use futures_util::stream::iter;
+    use rand::{thread_rng, Rng};
+
+    use crate::{
+        error::Error,
+        ff::Fp31,
+        protocol::{
+            basics::SecureMul,
+            context::{
+                dzkp_validator::{DZKPBaseField, DZKPValidator},
+                Context, UpgradableContext,
+            },
+            RecordId,
+        },
+        secret_sharing::{replicated::semi_honest::AdditiveShare as Replicated, IntoShares},
+        test_fixture::{join3v, Reconstruct, TestWorld},
+    };
+
+    // todo: remove impl and add support for BooleanTypes
+    impl DZKPBaseField for Fp31 {
+        type UnverifiedFieldValues = ();
+
+        fn convert(
+            _x_left: &BitArray<[u8; 32], Lsb0>,
+            _x_right: &BitArray<[u8; 32], Lsb0>,
+            _y_left: &BitArray<[u8; 32], Lsb0>,
+            _y_right: &BitArray<[u8; 32], Lsb0>,
+            _prss_left: &BitArray<[u8; 32], Lsb0>,
+            _prss_right: &BitArray<[u8; 32], Lsb0>,
+            _z_right: &BitArray<[u8; 32], Lsb0>,
+        ) {
+        }
+    }
+
+    /// test for testing `validated_seq_join`
+    /// similar to `complex_circuit` in `validator.rs`
+    #[tokio::test]
+    async fn complex_circuit_dzkp() -> Result<(), Error> {
+        const COUNT: usize = 100;
+        let world = TestWorld::default();
+        let context = world.malicious_contexts();
+        let mut rng = thread_rng();
+        let chunk_size: usize = 1 << rng.gen_range(1..10);
+
+        let mut original_inputs = Vec::with_capacity(COUNT);
+        for _ in 0..COUNT {
+            let x = rng.gen::<Fp31>();
+            original_inputs.push(x);
+        }
+        let shared_inputs: Vec<[Replicated<Fp31>; 3]> = original_inputs
+            .iter()
+            .map(|x| x.share_with(&mut rng))
+            .collect();
+        let h1_shares: Vec<Replicated<Fp31>> = shared_inputs.iter().map(|x| x[0].clone()).collect();
+        let h2_shares: Vec<Replicated<Fp31>> = shared_inputs.iter().map(|x| x[1].clone()).collect();
+        let h3_shares: Vec<Replicated<Fp31>> = shared_inputs.iter().map(|x| x[2].clone()).collect();
+
+        let futures = context
+            .into_iter()
+            .zip([h1_shares, h2_shares, h3_shares])
+            .map(|(ctx, input_shares)| async move {
+                let v = ctx.dzkp_validator(chunk_size);
```

Comment:
Is there some kind of assert statement later that validates that this DZKP validator ran and ran correctly?

Reply:
Yes, there is a custom implementation of the drop trait so if the validator is dropped and not every multiplication has been verified, it will cause a panic.

Reply:
Right now the backend is not integrated in the API yet. Once we have done that, I will add a test that will fail.